gap/attributes/isomorph.gi

```diff
@@ -171,50 +190,150 @@ end);
+  end;
+
+  out := List([1 .. 2 * n ^ 2 + 2 * n], x -> []);
+  colors := ListWithIdenticalEntries(n, 1);
+  Append(colors, ListWithIdenticalEntries(n ^ 2, 2));
+  Append(colors, ListWithIdenticalEntries(n ^ 2, 3));
+  Append(colors, ListWithIdenticalEntries(n, 2));
+
+  for x in [1 .. n] do
+    for y in [1 .. n] do
+      Add(out[Color1Node(x)], Color2Node(x, y));
+      if x = y then
+        Add(out[Color2Node(x, y)], Widget(x));
+      fi;
+      Add(out[Color2Node(x, y)], Color3Node(x, y));
+      Add(out[Color3Node(x, y)], M[x][y]);
+      for z in [1 .. n] do
+        if z <> x then
+          Add(out[Color2Node(x, y)], Color2Node(z, y));
+        fi;
+      od;
+    od;
+  od;
+  return [Digraph(out), colors];
+end;
+
 InstallMethod(IsomorphismSemigroups, "for semigroups",
 [IsSemigroup, IsSemigroup],
 function(S, T)
-  local pS, pT;
+  local invariants, map, DS, DT, p, inv;
 
+  # TODO more invariants
+  invariants := [IsFinite, IsSimpleSemigroup, IsZeroSimpleSemigroup, Size,
+                 NrLClasses, NrDClasses, NrRClasses, NrHClasses, NrIdempotents];
   if S = T then
     return MagmaIsomorphismByFunctionsNC(S, T, IdFunc, IdFunc);
   elif IsFinite(S) <> IsFinite(T) then
     return fail;
   elif not IsFinite(S) then
     TryNextMethod();
-  elif IsSimpleSemigroup(S) <> IsSimpleSemigroup(T) or
-      IsZeroSimpleSemigroup(S) <> IsZeroSimpleSemigroup(T) or
-      Size(S) <> Size(T) or NrRClasses(S) <> NrRClasses(T) or
-      NrDClasses(S) <> NrDClasses(T) or NrLClasses(S) <> NrLClasses(T) or
-      NrHClasses(S) <> NrHClasses(T) or NrIdempotents(S) <> NrIdempotents(T)
-      then
-    return fail;
-  fi;
-
-  if IsSimpleSemigroup(S) or IsZeroSimpleSemigroup(S)
+  elif IsTransformationSemigroup(T)
+      and HasIsomorphismTransformationSemigroup(S) then
+    map := IsomorphismTransformationSemigroup(S);
+    if T = Range(map) then
+      return map;
+    fi;
+  elif (IsSimpleSemigroup(S) and IsSimpleSemigroup(T))
+      or (IsZeroSimpleSemigroup(S) and IsZeroSimpleSemigroup(T))
       or (IsMonogenicSemigroup(S) and IsMonogenicSemigroup(T)) then
     return IsomorphismSemigroups(S, T);
+  elif not ForAll(invariants,
+                  func -> func(S) = func(T)) then
+    return fail;
   fi;
 
-  # Compare the partial orders of the D-classes
-  pS := DigraphReflexiveTransitiveClosure(Digraph(PartialOrderOfDClasses(S)));
-  pT := DigraphReflexiveTransitiveClosure(Digraph(PartialOrderOfDClasses(T)));
-  if not IsIsomorphicDigraph(pS, pT) then
+  DS := SEMIGROUPS.DigraphFromMultTable(S);
+  DT := SEMIGROUPS.DigraphFromMultTable(T);
+  p := IsomorphismDigraphs(DS[1], DT[1], DS[2], DT[2]);
+  if p = fail then
     return fail;
   fi;
+  p := RestrictedPerm(p, [1 .. Size(S)]);
+  map := x -> AsSSortedList(T)[PositionSorted(S, x) ^ p];
+  inv := x -> AsSSortedList(S)[PositionSorted(T, x) ^ (p ^ -1)];
+  return MagmaIsomorphismByFunctionsNC(S, T, map, inv);
+end);
+
+InstallMethod(AutomorphismGroup, "for a semigroup",
```

Comment:
I wonder if it's possible to just compute the automorphism group of the cayley digraph without edge colours, then wave some group theory magic wand to restrict the automorphisms found. That way you don't have to produce gigantic digraphs (though you do need a magic wand).

Reply:
I'm not sure, I tried something like this, but maybe didn't have a big enough wand. For example, for isomorphism testing, the semigroups with multiplication tables:
~~~~
1111
1111
1111
1112
~~~~
and 
~~~~
1111
1111
1111
1121
~~~~
are not isomorphic (or anti-isomorphic) and if you do `Digraph(MultiplicationTable())`, then the resulting digraphs are **equal** (according to the operator = for digraphs in GAP) not only isomorphic. I have no doubt that the method I've given here is a fallback and that there are better methods for some types of semigroups (see the comment about RZMS for example). Anyway, it's better than nothing, and I don't know any immediate ways to improve it.